### PR TITLE
Support for jr:instance/last-saved endpoint

### PR DESCRIPTION
**Pull request by eyelidlessness**
Addresses #137.

Moving my progress checklist over here since #258 will only address a subset of this issue.

- [x] Whenever a record is saved to storage save a copy with some __lastSaved key. This makes sure that even successfully submitted records may serve as a source for lastSaved.
  - [x] Offline
  - [x] Online
- [x] Populate a secondary instance with that record 
- [x] Resolve the jr://instance/last-saved URL
- [x] check that this feature is disabled if a form has encryption enabled.
- [x] check that it is not shown in record list
- [x] check that is is not submitted
- [x] check multiple form IDs
- [x] check it is not included in the export of local records

I'll add some additional notes/details inline once the PR is opened.

### CODE REVIEW

File: app/models/record-model.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * @see {@link https://getodk.github.io/xforms-spec/#instance}
+ * @see {ClientStore}
+ */
+
+/** Note: currently this module exists to provide a type definition for EnketoRecord
+ * in a place consistent with where other data model types are defined. This export
+ * is only present because an ESM module (if this is treated as one) must import
+ * or export _something_.
+ */
+export {};
```
Comment (eyelidlessness) [inline app/models/record-model.js:25]: This isn’t strictly necessary but it’s a habit I’ve developed to ease potential ESM transitions.

File: app/models/survey-model.js [hunk 2]
```diff
@@ -18,14 +18,74 @@ if ( process.env.NODE_ENV === 'test' ) {
...
+ * @property { string } [returnUrl]
+ * @property { string } [xslHash]
+ * @description
+ *   `SureveyObject` is Enketo's internal representation of an XForm, with some
+ *   additional properties representing resolved/deserialized external data.
+ *   This type definition captures the current state of "what is"—i.e. the full
+ *   known set of properties which may be added to a `SureveyObject` through
+ *   several data flow paths throught enketo-express. Some related resources,
+ *   notably those describing instances, are only populated in paths specific
+ *   to the interaction between a `SurveyObject` and those resources.
```
Comment (eyelidlessness) [inline app/models/survey-model.js:88]: 1. While this is, I believe, descriptive of the various states a `Survey` might be in at various points in the data flow, it’s obviously not what one would consider a survey in the sense it might be understood as a Form representation. Late in the process of this PR, @lognaturel discussed strategies for gradually decoupling things that are attached to data types this way. The first such decoupling is removal of an `htmlView` property, which you’ll see further down in the PR.

File: app/models/survey-model.js [hunk 3]
```diff
@@ -35,7 +95,7 @@ if ( process.env.NODE_ENV === 'test' ) {
  * @name get
  * @function
  * @param { string } id - Survey ID
- * @return {Promise<Error|SurveyObject>} Promise that resolves with a survey object
+ * @return {Promise<SurveyObject>} Promise that resolves with a survey object
```
Comment (eyelidlessness) [inline app/models/survey-model.js:98]: While I wish there were some way to express a Promise’s rejection types, this unfortunately wasn’t accurate and may be confusing at call sites.

File: public/js/src/enketo-webform.js [hunk 4]
```diff
@@ -43,8 +43,9 @@ if ( settings.offline ) {
         .catch( _showErrorOrAuthenticate );
 } else {
     console.log( 'App in online-only mode.' );
-    initTranslator( survey )
-        .then( connection.getFormParts )
+    store.init()
```
Comment (eyelidlessness) [inline public/js/src/enketo-webform.js:46]: This probably speaks to the question about support in #137, it’s odd to initialize `store` here, but it was necessary to enable that support. I think it would be unusual not to have this feature parity, but if this isn’t preferred (or if there’s a better way to handle it) I’m all ears.

Comment (eyelidlessness) [reply public/js/src/enketo-webform.js:46]: Also, it occurs to me that some changes here may also be relevant to `enketo-webform-edit.js`? Unclear as I’m unsure how I might access that, I was hesitant to make changes there without tests or manual validation.

File: public/js/src/module/connection.js [hunk 5]
```diff
@@ -44,26 +89,40 @@ function uploadRecord( record ) {
...
 
-            return results[ 0 ];
-        } );
+            result = results[ 0 ];
+
+            if ( options.isLastSaved ) {
+                return formCache.setLastSavedRecord( record.enketoId, record );
+            }
+        } )
+        .then( () => result );
```
Comment (eyelidlessness) [inline public/js/src/module/connection.js:118]: While this is clearly adding last-saved support, I also discovered adding initial test coverage that the original return Promise was always resolving `undefined` because these results were not actually being captured. That said it’s not clear why we would capture them all if only the first is relevant. Mostly I tried to stick close to the apparent intent.

File: public/js/src/module/connection.js [hunk 6]
```diff
@@ -87,7 +146,8 @@ function _uploadBatch( recordBatch ) {
         body: recordBatch.formData
     } )
         .then( response => {
-            const result = {
+            /** @type { UploadBatchResult } */
+            let result = {
```
Comment (eyelidlessness) [inline public/js/src/module/connection.js:150]: Besides adding types as I go to understand data flow, I’ve tried to also use `let` instead of `const` to identify where a reference is being treated as mutable. There are differing opinions on this, but this has become the most common idiom and can be helpful in understanding how state changes affect data flow.

File: public/js/src/module/controller-webform.js [hunk 7]
```diff
@@ -133,6 +155,16 @@ function _checkAutoSavedRecord() {
...
+/**
+ * @param {string} enketoId
+ */
+function _updateFormData( enketoId ) {
+    return formCache.get( { enketoId } )
+        .then( survey => {
+            formData.external = survey.externalData;
+        } );
+}
+
```
Comment (eyelidlessness) [inline public/js/src/module/controller-webform.js:167]: I really did not want to add more logic in this module without tests, but this was the only way I could find to ensure last-saved secondary instance references would update on save/form reset.

File: public/js/src/module/encryptor.js [hunk 8]
```diff
@@ -25,6 +34,49 @@ function isSupported() {
...
+ * @param {Survey} survey
+ * @return {Survey}
+ */
+function setEncryptionEnabled( survey ) {
+    return Object.defineProperty( survey, isEncryptionEnabledSymbol, {
+        configurable: false,
+        value: true,
+    } );
+}
+
```
Comment (eyelidlessness) [inline public/js/src/module/encryptor.js:79]: This whole set of changes was motivated by reducing/not adding to business logic coupling across concerns. It’s possibly over-engineered in general to hide implementation with `Symbol`, but I think it’s good in general for encapsulation… and especially around anything security related that might be subject to tampering (or even mistakes that have the same outcome as tampering).

File: public/js/src/module/form-cache.js [hunk 9]
```diff
@@ -25,21 +44,80 @@ function init( survey ) {
...
+ * @param {Survey} survey
+ * @return {Promise<void>}
+ */
 function remove( survey ) {
     return store.survey.remove( survey.enketoId );
 }
 
+/**
+ * @param {Survey} survey
+ * @return {PRomise<Survey>}
```
Comment (eyelidlessness) [inline public/js/src/module/form-cache.js:119]: Spotted this typo too late 😬. I’m on mobile right now so I’ll fix this and anything else I spot in the morning.

File: public/js/src/module/form-cache.js [hunk 10]
```diff
@@ -102,11 +184,11 @@ function _setUpdateIntervals( survey ) {
     // that open the form right after the XForm update.
     setTimeout( () => {
         _updateCache( survey );
-    }, 3 * 1000 );
+    }, CACHE_UPDATE_INITIAL_DELAY );
```
Comment (eyelidlessness) [inline public/js/src/module/form-cache.js:187]: These delays were extracted to constants both for clarity and to aid testing with fake timers.

File: public/js/src/module/records-queue.js [hunk 11]
```diff
@@ -11,16 +11,20 @@ import settings from './settings';
...
 import $ from 'jquery';
+import formCache from './form-cache';
 
 let $exportButton;
 let $uploadButton;
 let $recordList;
 let $queueNumber;
 let uploadProgress;
 let finalRecordPresent;
-const autoSaveKey = `__autoSave_${settings.enketoId}`;
```
Comment (eyelidlessness) [inline public/js/src/module/records-queue.js:21]: This was moved to a function body primarily to aid testing, but also leaves us in a better position to reduce reliance on global state in the future.

File: public/js/src/module/store.js [hunk 12]
```diff
@@ -296,47 +306,49 @@ const surveyStore = {
...
-                    theme: survey.theme,
-                    // Note: if survey.resources = undefined/null, do not store empty array
-                    // as it means there are no resources to store (and load) - they may be added later by form-cache.js
-                    resources: survey.resources ? resourceKeys : undefined,
-                    binaryDefaults: survey.binaryDefaults ? binaryDefaultKeys : undefined,
-                    maxSize: survey.maxSize,
-                    externalData: survey.externalData,
-                    branding: survey.branding
-                } );
+                return server.surveys.update( update );
```
Comment (eyelidlessness) [inline public/js/src/module/store.js:322]: This set of changes was primarily about minimizing the difference and potential drift between create/update logic, as it was quite confusing in testing why `lastSavedRecord` was being set once then dropped thereafter.

File: public/js/src/module/store.js [hunk 13]
```diff
@@ -296,47 +306,49 @@ const surveyStore = {
...
-                    .then( () => // resolving with original survey (not the array returned by server.surveys.update)
-                        survey )
-                    .then( _transformExternalDataToXmlDoc );
+                    .then( resources => {
+                        if ( survey.resources ) {
+                            result.resources = resources.filter( resource => resource != null );
+                        }
+
+                        return _deserializeSurvey( result );
+                    } );
```
Comment (eyelidlessness) [inline public/js/src/module/store.js:344]: This was also about minimizing the difference between create/update, but also and more importantly about being able to validate that the actual update produces the expected result, where returning the incoming parameter might mask problems.

File: public/js/src/module/store.js [hunk 14]
```diff
@@ -667,32 +680,36 @@ function _firstItemOnly( results ) {
...
-            return instance;
-        } );
-    }
+                return instance;
+            } );
+        }
 
-    return survey;
+        return encryptor.deserializeEncryptedSurvey( survey );
+    }
```
Comment (eyelidlessness) [inline public/js/src/module/store.js:712]: In hindsight, I wish I had defined an actual `Survey` class which handles this (as well as documenting its actual type). I had also thought it might be possible to use a method similar to `toJSON`, but IndexedDB depends on serialization rules without an equivalent method. For future consideration, it might be good for our wrapper to provide a generalized (optional) mechanism for de/serialization of the types it handles.

File: test/client/form-cache.spec.js [hunk 15]
```diff
@@ -86,40 +194,179 @@ describe( 'Client Form Cache', () => {
 
     } );
 
-    /*
-    describe( 'in cached state', function() {
-        
-        it( 'initializes succesfully', function( done ) {
+    describe( 'last-saved records', () => {
```
Comment (eyelidlessness) [inline test/client/form-cache.spec.js:197]: The diff algorithm really botched this. I highly recommend just looking at the before/after if there’s confusion about what changed.

File: public/js/src/module/connection.js [hunk 16]
```diff
@@ -370,31 +454,55 @@ function _encodeFormData( data ){
         .join( '&' );
 }
 
+/**
+ * @param {Document} model
+ * @param {EnketoRecord} [lastSavedRecord]
+ * @return {XMLDocument}
+ */
+function _getLastSavedInstance( model, lastSavedRecord ) {
+    if ( lastSavedRecord == null ) {
```
Comment (lognaturel) [inline public/js/src/module/connection.js:463]: We should document this fallback behavior.

File: test/client/connection.spec.js [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+
+                    return submission.text();
+                } )
+                .then( submission => {
+                    expect( submission ).to.equal( record.xml );
+                } )
+                .then( done, done );
+        } );
+
+        describe( 'last-saved records', () => {
```
Comment (lognaturel) [inline test/client/connection.spec.js:132]: Some of these feel like they'd also be good candidates for pulling out to a common feature test place. But this might really be a matter of taste.

File: test/client/connection.spec.js [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+                        return formCache.getLastSavedRecord( enketoId );
+                    } )
+                    .then( ( record ) => {
+                        expect( record ).to.deep.equal( originalRecord );
+                    } )
+                    .then( done, done );
+            } );
+
+            // Note: this allows records-queue to continue to upload when transitioning
+            // from offline to online.
```
Comment (lognaturel) [inline test/client/connection.spec.js:177]: Ah, helpful, thanks.

**Review by lognaturel — COMMENTED** [review]
Great to see all these tests!! Do you know what's up with the three that are failing on CI? 

A few questions inline. As I was going through I looked for areas of risk or things I really didn't understand and flagged those. It's a lot to digest so I'm going to let it all marinate in my brain a bit now that I know what the surface area is and do another pass again looking for risk and edge cases.

**Comment by eyelidlessness** [comment]
Thanks for all the feedback! I’ll get to more in a bit when I’m back at my desk but this one I can address now:

> Great to see all these tests!! Do you know what's up with the three that are failing on CI?

Funny enough, I was right in the middle of answering this at the same time but got interrupted with a phone call. This is yet another case where the version of Chrome in CI is way out of date (which increasingly feels like a risk beyond just dev/CI differences… who knows what API changes are lurking as several browser vendors have recently begun actually removing/restricting some functionality 😬).

Some of these new tests cover code paths into previously untested code that depends on `AbortController`, which isn’t available in the CI environment’s latest version of Chrome. Noting just to have a clear picture of how out of date that is: `AbortController` shipped [April 2018](https://caniuse.com/abortcontroller). 🤯

Given that previous attempts to address this proved too big and out of scope for this work, I think the prudent thing to do would be to:

- Add a polyfill, only in test
- Add a comment noting that this is meant to be removed once the CI issue is resolved
- Consider adopting an [ESLint rule](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/expiring-todo-comments.md) to expire TODOs so this doesn’t just linger forever
- Discuss prioritizing the CI changes necessary to get more up to date on Chrome

Related note: I recently got curious about Enketo’s browser support strategy and I was pleasantly surprised to find it’s relatively aggressive about supporting modern/evergreen versions (excepting Safari because of course it has to be an exception). Looooong term I think it’s a good idea to set ourselves up to be as confident as possible in those claims by setting up version matrix testing in CI.

File: public/js/src/module/store.js [hunk 19]
```diff
@@ -304,32 +307,42 @@ const surveyStore = {
...
+
+                // Note: if survey.resources = undefined/null, do not store empty array
+                // as it means there are no resources to store (and load) - they may be added later by form-cache.js
+                if ( survey.resources ) {
+                    update.resources = resourceKeys;
+                }
+
+                if ( survey.binaryDefaults ) {
+                    update.resources = binaryDefaultKeys;
+                }
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/store.js:319]: Oof, good copypasta catch! This change is primarily to avoid adding unnecessary properties now that `store.survey.update` returns the updated IndexedDB result rather than the incoming `survey` argument (these properties broke a lot of tests unnecessarily and would have introduced more noise), and partly stylistic. Of course not everyone agrees, but it's _generally_ idiomatic in JS to treat `null` as an explicit value and `undefined` as an absence of a property or return value. It's also generally idiomatic to avoid referencing `undefined` as a value because ~~everything is terrible~~ [it's not a reserved keyword](https://eslint.org/docs/rules/no-undefined).

File: test/client/records-queue.spec.js [hunk 20]
```diff
@@ -0,0 +1,526 @@
...
+                } )
+                .then( record => {
+                    Object.entries( update ).forEach( ( [ key, value ] ) => {
+                        expect( record[key] ).to.deep.equal( value );
+                    } );
+                } )
+                .then( done, done );
+        } );
+
+        it( 'creates a last-saved record when creating a record', done => {
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply test/client/records-queue.spec.js:229]: I like this suggestion (and similar ones) a lot! And it's very much in the BDD spirit of the test framework APIs. It's generally more common in JS/TS to pair test modules to a module of the same name, which can make discovery a little easier, but I really do prefer the feature-centered approach. I'll look at that in the morning.

File: test/client/records-queue.spec.js [hunk 21]
```diff
@@ -0,0 +1,526 @@
...
+                .then( records => {
+                    expect( Array.isArray( records ) ).to.equal( true );
+                    expect( records.length ).to.equal( 0 );
+                } )
+                .then( done, done );
+        } );
+    } );
+
+    describe( 'Uploading records', () => {
+        it( 'uploads queued records, excluding auto-saved/last-saved and draft records', done => {
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply test/client/records-queue.spec.js:472]: I'd originally kept this in as it was part of the criteria in #137 and felt like a good regression test in case we decide to change the approach, but I agree it feels like irrelevant noise now.

File: public/js/src/module/store.js [hunk 22]
```diff
@@ -668,32 +681,36 @@ function _firstItemOnly( results ) {
...
+    }
 }
 
-function _transformExternalDataToXmlDoc( survey ) {
-    if ( survey && survey.externalData ) {
-        survey.externalData = survey.externalData.map( instance => {
-            if ( typeof instance.xml === 'string' ) {
-                instance.xml = parser.parseFromString( instance.xml, 'text/xml' );
-            }
+function _deserializeSurvey( survey ) {
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/store.js:699]: IndexedDB requires any stored value to be serialized in full, and while many data types are serializable some unfortunately aren't. I'm having trouble finding the original documentation I referenced but [this MDN doc](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) and [the W3C spec](https://html.spec.whatwg.org/multipage/structured-data.html#safe-passing-of-structured-data) go into some detail here.

`Symbol` (the mechanism used in `encryptor` to mark a Survey as having encryption enabled) is notably _not_ serializable. So this additional step to serialization transforms `Symbol('isEncryptionEnabled')` to a plain string property `isEncryptionEnabled` for storage, then back to the symbol when retrieved in the runtime.

My feeling was that it was safer to have a single point for de/serialization of Survey than add another field-specific set of private functions which would _always_ be paired with the existing one.

[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/store.js:699]: In the fullness of time, I would prefer to separate concerns like this entirely with clearly delineated runtime (and type safe) validators/transformers not unlike [io-ts](https://github.com/gcanti/io-ts) or [Runtypes](https://github.com/pelotom/runtypes).

File: public/js/src/module/connection.js [hunk 23]
```diff
@@ -370,31 +454,55 @@ function _encodeFormData( data ){
         .join( '&' );
 }
 
+/**
+ * @param {Document} model
+ * @param {EnketoRecord} [lastSavedRecord]
+ * @return {XMLDocument}
+ */
+function _getLastSavedInstance( model, lastSavedRecord ) {
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/connection.js:462]: That was my thinking too (and the most recent source of hemming and hawing around circular dependencies).

File: public/js/src/module/connection.js (hunk 16 above)
Comment (eyelidlessness) [reply public/js/src/module/connection.js:463]: Oh, good call! Do you mean in the JSDoc, or also additional documentation?

**Comment by eyelidlessness** [comment]
@lognaturel I answered some questions and made a few of the requested changes (those with 👍 reactions). Anything I haven't acknowledged, I'm in agreement and I'll work on the remaining changes tomorrow.

File: test/client/records-queue.spec.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (lognaturel) [reply test/client/records-queue.spec.js:229]: I'm glad to see the concept resonates. Of course, the change doesn't necessarily have to happen in this PR unless it's straightforward and does end up adding clarity. I mostly want it out there as something to consider as more tests get added.

> which can make discovery a little easier

Depending on which "direction" you're coming from! Like if you see some last-saved behavior in the form cache or just know about the feature and want to see corresponding tests, these records-queue tests might be hard to find. The biggest challenge I have with a purely module-based structure is that it's hard to reason about conceptual coverage (not line coverage) because an edge case could be tested in a number of places.

File: public/js/src/module/connection.js (hunk 16 above)
Comment (lognaturel) [reply public/js/src/module/connection.js:463]: I meant in user-facing docs but actually I think it's non-obvious enough that it's good to have in the JSDoc briefly as well.

Comment (eyelidlessness) [reply public/js/src/module/connection.js:463]: Turns out those docs are _generated by_ JSDoc, but only from `/app` (the server side). Fortunately that's where the Survey type is documented, so I've added a note there, as well as here.

File: public/js/src/module/store.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (lognaturel) [reply public/js/src/module/store.js:699]: I see, that all makes sense. I'm still not a huge fan of the name because it feels like it's not clear what is being deserialized or why but don't have an alternative to propose at this time. Maybe a quick comment here would help provide context?

**Comment by lognaturel** [comment]
I've done another pass through and nothing else jumps out at me. I've tried it and it works really well! 🤩

There is a regression around edits. I now always see `Cannot read property 'surveys' of undefined` which suggests to me that it's related to the form cache.

As I've been thinking more about usage and user impact, I think we should look into only saving the last-saved record if the form asks for it and/or expiring last-saved records as part of the initial release. Only doing the save if the form asks for it could be relatively straightforward depending on when `externalData` gets populated. 

I think saving the last-saved record is generally ok in an enumerator-mediated context though even then there are shared device cases in which it wouldn't be great. It feels pretty bad in a direct respondent case, though. Imagine someone filling out a health-related survey on a public computer. They would not expect their data to stay in indexdb for ever.

In summary, I think what remains is:
- [ ] Fix a few more small things from feedback
- [ ] Fix edit regression
- [ ] Possibly introduce a feature test concept (though maybe best to do separately)
- [ ] Explore only storing last-saved record if the form def asks for it (could be separate PR but probably wouldn't want to release without something like this or expiration)
- [ ] Fix failing tests

All that said, I don't think there are any major changes coming and it would be great to get @MartijnR's eyes on it sooner than later.

**Comment by eyelidlessness** [comment]
> Investigate possible regression in offline mode where a draft is said to exist when it shouldn't

Turns out this is not a regression, I see the same behavior on `master`. Filed an issue in #275.

**Comment by lognaturel** [comment]
Looking great from my end, thanks! I really like how the feature tests are looking. I feel like they really help document the totality of the `last-saved` and `encryption` behavior.

File: public/js/src/module/records-queue.js [hunk 24]
```diff
@@ -356,10 +398,8 @@ function _updateRecordList() {
     // rebuild the list
     return store.record.getAll( settings.enketoId )
```
Comment (MartijnR) [inline public/js/src/module/records-queue.js:399]: The comment for getDisplaybleRecordList implies you perhaps meant to call that function, or did I misunderstand?

File: public/js/src/module/store.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (MartijnR) [reply public/js/src/module/store.js:719]: I'm not quite convinced that the Symbol (for setting isEncryptionEnabled) adds meaningful security but I may very well be overlooking something. My thinking is that if somebody wants to prevent encryption from happening they'd already have access to the not-yet-encrypted instance. Wouldn't changing this setting be the same level of effort as simply accessing the model before the record is encrypted?

**Review by MartijnR — COMMENTED** [review]
Wow, so much work. Thanks!!

It took me a long time to read through the code in my spare moments, and I just wanted to send this already, even though I'm still planning to run this myself for some real-life testing. I might get interrupted again, but I thought it was taking too long for me to at least give some feedback. 

I'll still ping you to talk through the comments, so it would be fine with me to wait for that instead of responding inline.

File: public/js/src/module/connection.js [hunk 25]
```diff
@@ -356,28 +454,58 @@ function _encodeFormData( data ){
...
+ * Gets a survey's last-saved record data as an `XMLDocument`, defaulting to
+ * the survey's model when no last-saved record exists.
+ *
+ * @param {Document} model
+ * @param {EnketoRecord} [lastSavedRecord]
+ * @return {XMLDocument}
+ */
+function _getLastSavedInstance( model, lastSavedRecord ) {
+    if ( lastSavedRecord == null ) {
+        const modelDefault = model.querySelector( 'model > instance > *' ).cloneNode( true );
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/connection.js:471]: Yep. This felt like a bit of a hack, but it was a relatively straightforward path to ensuring forms with last-saved secondary instances behave normally before a last-saved record is available.

File: public/js/src/module/encryptor.js [hunk 26]
```diff
@@ -25,6 +34,49 @@ function isSupported() {
...
+ */
+function isEncryptionEnabled( survey ) {
+    return Boolean( survey[isEncryptionEnabledSymbol] );
+}
+
+function serializeEncryptedSurvey( survey ) {
+    if ( isEncryptionEnabled( survey ) ) {
+        return Object.assign( {}, survey, {
+            isEncryptionEnabled: true,
+        } );
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/encryptor.js:59]: In general I tend to avoid mutating function parameters unless there's a compelling reason to do so, as it tends to make it easier to reason about state/data flow.

File: public/js/src/module/form-cache.js [hunk 27]
```diff
@@ -25,21 +44,80 @@ function init( survey ) {
...
+    return store.survey.get( enketoId )
+        .then( survey => {
+            if ( encryptor.isEncryptionEnabled( survey ) ) {
+                return Promise.resolve( survey );
+            }
+
+            if ( Array.isArray( survey.externalData ) ) {
+                const { xml } = lastSavedRecord;
+
+                survey.externalData.forEach( item => {
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/form-cache.js:106]: It actually just didn't occur to me to use `find` here! You're right about the performance impact, but I think using `find` may actually make this easier to understand. And it would also be a better place to ensure `lastSavedRecord` isn't stored when the form doesn't reference it.

[the remark this answers is not in the document]
Comment (eyelidlessness) [reply public/js/src/module/form-cache.js:106]: As far as other usage of `forEach` I think they're mostly in tests, where the intent was to test deep equality over a subset of properties. Although I see lodash is already in use in the project, so I may switch those to `_.pick` calls if you think that would make those cases easier to understand?

I tend to avoid lodash except when it's already present and improves code readability, but this may be one of those cases. Not because I'm one of those "you don't need a utility library" purists, but because lodash has quite a few footguns (e.g. confusing argument order, surprising implicit behavior when some arguments are omitted, unexpected mutation).

File: app/models/survey-model.js [hunk 28]
```diff
@@ -18,14 +18,77 @@ if ( process.env.NODE_ENV === 'test' ) {
     client.select( 15 );
 }
 
+/**
+ * @typedef {import('./account-model').AccountObj} AccountObj
```
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply app/models/survey-model.js:22]: Oof, yes that does error. That's going to make the gradual typing process a lot more challenging. It looks like [jsdoc-ts-utils](https://github.com/homer0/jsdoc-ts-utils) may help address this. I'll give it a quick look, as I'm highly motivated to keep improving the type coverage.

File: public/js/src/module/connection.js (hunk 5 above)
Comment (eyelidlessness) [reply public/js/src/module/connection.js:118]: I'm not sure what you mean. But to clarify:

1. The original behavior in the `reduce` callback didn't accumulate results, it just stepped through each batch and returned the final `Promise<UploadBatchResult>`.
2. The next `then` refers to `results` and `results[ 0 ]` so it appeared the intent/expectation was that the `reduce` _was_ accumulating results, so that the _first_ batch's `UploadBatchResult` could be returned. Instead, `results[ 0 ]` was being coerced to `results[ '0' ]`, where the `UploadBatchResult` property did not have a `'0'` property, so it would always resolve to `undefined`.
3. While I was making the change here to support last-saved, I also addressed 1 & 2 by accumulating results as `Promise<Array<UploadBatchResult>>`, then resolving the first one, as that did appear to be the original intent.

File: public/js/src/module/records-queue.js (hunk 24 above)
Comment (eyelidlessness) [reply public/js/src/module/records-queue.js:399]: Good catch, yes!

File: app/models/survey-model.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (eyelidlessness) [reply app/models/survey-model.js:22]: With a few minor changes that worked!

File: public/js/src/module/connection.js (hunk 5 above)
Comment (MartijnR) [reply public/js/src/module/connection.js:118]: Thanks. That seems great.

File: public/js/src/module/form-cache.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (MartijnR) [reply public/js/src/module/form-cache.js:106]: > I tend to avoid lodash 

Yes, I do too. The existing use of lodash is probably no longer necessary either.
